Add support for sitemap.xml

Zineb-Ada opened the pull request:
What we've done here is create a program that generates the sitemap.xml for the entire `data` files.
The PR is in progress, I will add another one for `packages`.
This PR responds to issue #26

Zineb-Ada committed:
Add support for sitemap.xml

cuihtlauac reviewed (changes requested):
Files in `playground/asset` should not be affected by this PR. You probably need to rebase and force push.

Zineb-Ada committed:
Generate sitemap.xml by dream

cuihtlauac commented on tool/sitemap/sitemap.ml:
```diff
@@ -0,0 +1,84 @@
+    :: list_of_urls_lists
+  in
+  let list_of_complete_urls =
+    List.map (fun url -> cu url) (List.concat list_of_urls) @ list_of_v2_urls
```
```suggestion
    List.map cu (List.concat list_of_urls) @ list_of_v2_urls
```

cuihtlauac commented on tool/sitemap/sitemap.ml:
```diff
@@ -0,0 +1,84 @@
+      Url.problems;
+      Url.installer;
+    ]
+    :: list_of_urls_lists
```
```suggestion
    :: [ success_story_urls; release_urls; workshop_urls; news_urls; tutorial_urls ]
```

cuihtlauac commented on tool/sitemap/sitemap.ml:
```diff
@@ -0,0 +1,84 @@
+  [ Url.v2; Url.manual_with_version "5.0.0"; Url.api_with_version "5.0.0" ]
+
+let list_of_urls_lists =
+  [ success_story_urls; release_urls; workshop_urls; news_urls; tutorial_urls ]
```
```suggestion
```

Zineb-Ada committed:
make package documentation link more obvious (#828)

* make documentation link more obvious

* minor style improvement

Co-Authored-By: Sabine Schmaltz <sabine@tarides.com>

cuihtlauac commented on src/ocamlorg_web/lib/router.ml:
```diff
@@ -1,4 +1,5 @@
 module Url = Ocamlorg_frontend.Url
+module Sitemap = Sitemap
```
```suggestion
```

cuihtlauac commented on src/ocamlorg_frontend/url.ml:
```diff
@@ -11,6 +11,7 @@ let package_with_version ?hash v version =
 let package_doc ?hash ?(page = "index.html") v version =
   with_hash hash ^ "/" ^ v ^ "/" ^ version ^ "/doc/" ^ page
 
+let sitemap = "/sitemap"
```
```suggestion
let sitemap = "/sitemap.xml"
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,81 @@
+module Url = Ocamlorg_frontend.Url
```
```suggestion
```

cuihtlauac reviewed (changes requested):
In `sitemap.ml` function `cu` must be renamed into something meaningful

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,82 @@
+    | [] -> sitemap_body_list
+    | url :: t ->
+        sitemap_body t
+          (Printf.sprintf {|
```
```suggestion
  let f url = Printf.sprintf {|
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,82 @@
+<url>
+  <loc>%s</loc>
+</url>|} url
+          :: sitemap_body_list)
```
```suggestion
</url>|} url in
  let sitemap_body = List.map f
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,82 @@
+          :: sitemap_body_list)
+  in
+  let sitemap_list =
+    {|</urlset>|} :: sitemap_body list_of_complete_urls sitemap_list
```
```suggestion
    sitemap_list @ sitemap_body list_of_complete_urls @ [ {|</urlset>|} ]
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,82 @@
+  let sitemap_list =
+    {|</urlset>|} :: sitemap_body list_of_complete_urls sitemap_list
+  in
+  Lwt_seq.of_list (List.rev sitemap_list)
```
```suggestion
  Lwt_seq.of_list sitemap_list
```

Zineb-Ada committed:
Apply suggestions from @Cuihtlauac code review

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

Zineb-Ada committed:
Convert lists of urls to sequences and change names of functions

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,83 @@
+        (Ood.Workshop.all, fun r -> to_url @@ Url.workshop r.Ood.Workshop.slug);
+      Urlable (Ood.News.all, fun r -> to_url @@ Url.news_post r.Ood.News.slug);
+      Urlable
+        (Ood.Tutorial.all, fun r -> to_url @@ Url.tutorial r.Ood.Tutorial.slug);
```
```suggestion
        ( Success_story.all,
          fun r -> to_url @@ Url.success_story r.Success_story.slug );
      Urlable (Release.all, fun r -> to_url @@ Url.release r.Release.version);
      Urlable (Workshop.all, fun r -> to_url @@ Url.workshop r.Workshop.slug);
      Urlable (News.all, fun r -> to_url @@ Url.news_post r.News.slug);
      Urlable (Tutorial.all, fun r -> to_url @@ Url.tutorial r.Tutorial.slug);
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,83 @@
+<urlset
+      xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
+      xmlns:image="http://www.google.com/schemas/sitemap-image/1.1"
+      xmlns:xhtml="http://www.w3.org/1999/xhtml">
```
```suggestion
  xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
  xmlns:image="http://www.google.com/schemas/sitemap-image/1.1"
  xmlns:xhtml="http://www.w3.org/1999/xhtml">
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,83 @@
+  in
+  Lwt_seq.of_seq
+    (Seq.concat (List.to_seq [ header; urlset; Seq.return {|
+</urlset>|} ]))
```
```suggestion
</urlset>
|} ]))
```

cuihtlauac reviewed (changes requested):
Minor cleanup changes

sabine reviewed (changes requested):
I have two points here:

1. Sitemaps should not contain URLs from subdomains (https://webmasters.stackexchange.com/questions/82687/sitemaps-one-per-subdomain-or-one-for-the-base-domain)
2. Sitemaps should not contain URLs that return a non-200 status code (e.g. https://ocaml.org/releases/latest/manual.html, https://ocaml.org/releases/latest/api/index.html

I would recommend to replace both these types of URLs by comments reminding us to add these locations to the sitemap, as soon as they are both URLs on ocaml.org and return a 200 status code.

The code is obvious enough as is, so I wouldn't mind to merge this.

A follow-up improvement, since the data is all coming from compile-time sources, would be to emit a `sitemap.xml` static file and serving that, instead of dynamically rendering the sitemap on every access.

Zineb-Ada committed:
Apply suggestions from @Cuihtlauac code review 2

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

Zineb-Ada committed:
Delete subdomains URLs and URLs returning status code other than 200

cuihtlauac commented on src/ocamlorg_web/lib/handler.ml:
```diff
@@ -619,3 +619,11 @@ let package_doc t kind req =
+let sitemap _request =
+  let open Lwt.Syntax in
+  Dream.stream
+    ~headers:[ ("Content-Type", "application/octet-stream") ]
```
```suggestion
    ~headers:[ ("Content-Type", "application/xml; charset=utf-8") ]
```

Zineb-Ada committed:
Apply suggestions from @Cuihtlauac code review 3

Co-authored-by: Cuihtlauac Alvarado <cuihtlauac@users.noreply.github.com>

sabine committed:
improve authors/maintainers display on package overview (#1001)

sabine committed:
refactor package urls, routes, and handlers (#999)

* refactor package urls, routes, and handlers
* move package redirects to redirection.ml

sabine committed:
Unify package overview and documentation layout (#1015)

* unified package overview/docs layout

Rearranges package overview and documentation such that:

* package_layout.eml defines the two sidebars and the
  content area
* there is a navigation element to switch between
  Overview (About) and Documentation (Docs)
* there is a placeholder element for the upcoming
  in-package search (to remind us how important
  this feature is and to show users that
  we have a plan where it goes)

Consequences:

* package overview page now has a collapsing sidebar
  with a button to slide it in on small screens
* package documentation now has a tablet (md) layout
  that shows the sidebar on-screen, instead of collapsed

* add text-sm which was lost during unification

* "Overview" instead of "About" since we have the space

christinerose commented on HACKING.md:
```diff
@@ -88,14 +88,25 @@ The deployment pipeline is managed in <https://github.com/ocurrent/ocurrent-depl
 
+### Managing dependencies
+
+ocaml.org is using an Opam switch which is local and bound to a pinned commit in opam-repository. This is intended to protect the build from upstream regressions. The Opam repository is specified in three (3) places:
```
```suggestion
OCaml.org is using an opam switch that is local and bound to a pinned commit in `opam-repository`. This is intended to protect the build from upstream regressions. The opam repository is specified in three (3) places:
```

christinerose commented on HACKING.md:
```diff
@@ -88,14 +88,25 @@ The deployment pipeline is managed in <https://github.com/ocurrent/ocurrent-depl
 
 With the docker container running, you can visit the site at <http://localhost:8080/>.
 
+### Managing dependencies
```
```suggestion
### Managing Dependencies
```

christinerose commented on HACKING.md:
```diff
@@ -88,14 +88,25 @@ The deployment pipeline is managed in <https://github.com/ocurrent/ocurrent-depl
 
 This will build the docker image and run a docker container with the port `8080` mapped to the HTTP server.
 
 With the docker container running, you can visit the site at <http://localhost:8080/>.
```
```suggestion
With the Docker container running, you can visit the site at <http://localhost:8080/>.
```

christinerose commented on HACKING.md:
```diff
@@ -88,14 +88,25 @@ The deployment pipeline is managed in <https://github.com/ocurrent/ocurrent-depl
+
+When bringing up ocaml.org to a newer pin, the commit hash found it those files must be changed all at once.
+
 ## Repository structure
```
```suggestion
## Repository Structure
```

christinerose reviewed (approved):
capitalization mostly.... will review entire Hacking doc once this PR is merged.
Approving for your convenience, but it would be great to make these changes!

Zineb-Ada committed:
Update PR

Zineb-Ada committed:
formating

Zineb-Ada committed:
do not touch playground asset

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,75 @@
+let tag = Printf.sprintf {|
+<url>
+  <loc>%s</loc>
+</url>|}
```
```suggestion
<url><loc>%s</loc></url>|}
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,75 @@
+  Lwt_seq.of_seq
+    (Seq.concat (List.to_seq [ header; urlset; Seq.return {|
+</urlset>
+|} ]))
```
```suggestion
    (Seq.concat (List.to_seq [ header; urlset; Seq.return "\n</urlset>\n" ]))
```

Zineb-Ada committed:
Apply suggestions from @cuihtlauac code review 4

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,72 @@
+      Urlable (Release.all, fun r -> to_url @@ Url.release r.Release.version);
+      Urlable (Workshop.all, fun r -> to_url @@ Url.workshop r.Workshop.slug);
+      Urlable (News.all, fun r -> to_url @@ Url.news_post r.News.slug);
+      Urlable (Tutorial.all, fun r -> to_url @@ Url.tutorial r.Tutorial.slug);
```
```suggestion
      Success_story.(Urlable (all, fun r -> to_url @@ Url.success_story r.slug));
      Release.(Urlable (all, fun r -> to_url @@ Url.release r.version));
      Workshop.(Urlable (all, fun r -> to_url @@ Url.workshop r.slug));
      News.(Urlable (News.all, fun r -> to_url @@ Url.news_post r.slug));
      Tutorial.(Urlable (all, fun r -> to_url @@ Url.tutorial r.slug));
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,72 @@
+    ]
+
+let tag = Printf.sprintf {|
+<url><loc>%s</loc></url>|}
```
```suggestion
let tag = Printf.sprintf "\n<url><loc>%s</loc></url>"
```

Zineb-Ada committed:
Apply suggestions from @cuihtlauac code review 5

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,69 @@
+  Seq.concat_map
+    (function
+      | Urlable (all, show) -> Seq.map (fun s -> tag (show s)) (List.to_seq all))
+    urlables
```
```suggestion
let urlset (Urlable (all, show)) =
  Seq.map (fun s -> tag (show s)) (List.to_seq all)
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,69 @@
+  let header =
+    List.to_seq
+      [
+        {|<?xml version="1.0" encoding="utf-8"?>
```
```suggestion
    {|<?xml version="1.0" encoding="utf-8"?>
```

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,69 @@
+  xmlns:image="http://www.google.com/schemas/sitemap-image/1.1"
+  xmlns:xhtml="http://www.w3.org/1999/xhtml">
+|};
+      ]
```
```suggestion
```

Zineb-Ada replied:
I have to leave `|}` to close the string

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,69 @@
+      ]
+  in
+  Lwt_seq.of_seq
+    (Seq.concat (List.to_seq [ header; urlset; Seq.return "\n</urlset>\n" ]))
```
```suggestion
    (Seq.concat
       (List.to_seq
          [
            Seq.return header;
            Seq.concat_map urlset urlables;
            Seq.return "\n</urlset>\n";
          ]))
```

Zineb-Ada committed:
Apply suggestions from @cuihtlauac code review 6

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,68 @@
+            Seq.return header;
+            Seq.concat_map urlset urlables;
+            Seq.return "\n</urlset>\n";
+          ]))
```
```suggestion
    Seq.(
      concat
        (List.to_seq
           [ return header; concat_map urlset urlables; return "\n</urlset>\n" ]))
```

Zineb-Ada committed:
Apply suggestions from @cuihtlauac code review 7

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,65 @@
+      Tutorial.(Urlable (all, fun r -> to_url @@ Url.tutorial r.slug));
+    ]
+
+let tag = Printf.sprintf "\n<url><loc>%s</loc></url>"
```
```suggestion
let to_url u = "\n<url><loc>https://ocaml.org" ^ u ^ "</loc></url>"
```

Zineb-Ada replied:
I deleted line 34 and replaced it with `let to_url u = "\n<url><loc>https://ocaml.org" ^ u ^ "</loc></url>"`

cuihtlauac commented on src/ocamlorg_web/lib/sitemap.ml:
```diff
@@ -0,0 +1,65 @@
+let tag = Printf.sprintf "\n<url><loc>%s</loc></url>"
+
+let urlset (Urlable (all, show)) =
+  Seq.map (fun s -> tag (show s)) (List.to_seq all)
```
```suggestion
let urlset (Urlable (all, show)) = Seq.map show (List.to_seq all)
```

cuihtlauac reviewed (changes requested):
Sorry, yet another pedantic set of changes

Zineb-Ada committed:
Apply suggestions from @cuihtlauac code review 8
